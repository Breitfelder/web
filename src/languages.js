export default {
  python: {
    name: 'Python',
    url: 'https://github.com/bblfsh/python-driver',
    mode: 'python'
  },
  java: {
    name: 'Java',
    url: 'https://github.com/bblfsh/java-driver',
    mode: 'text/x-java'
  }
};
