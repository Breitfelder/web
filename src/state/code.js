import * as api from '../services/api';
import { set as languageSet } from './languages';
import { setUastQuery } from './options';
import { add as errorsAdd, clear as errorsClear } from './errors';
import { updateIfNeeded as versionUpdateIfNeeded } from './versions';
import { updateIfNeeded as driversUpdateIfNeeded } from './languages';

export const initialState = {
  filename: undefined,
  code: null,
  uastJson: null,

  dirty: false,
  parsing: false,
};

export const SET = 'bblfsh/code/SET';
export const PARSE = 'bblfsh/code/PARSE';
export const PARSE_FAILED = 'bblfsh/code/PARSE_FAILED';
export const SET_AST = 'bblfsh/code/SET_AST';
export const CHANGE = 'bblfsh/code/CHANGE';

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case SET:
      return {
        ...reducer(state, {
          type: SET_AST,
          ast: action.ast,
        }),
        filename: action.filename,
        code: action.code,
        markRange: null,
      };
    case PARSE:
      return {
        ...state,
        uastJson: null,
        posIndex: null,
        parsing: true,
      };
    case PARSE_FAILED:
      return {
        ...state,
        parsing: false,
      };
    case SET_AST:
      return {
        ...state,
        parsing: false,
        dirty: false,
        uastJson: action.ast,
      };
    case CHANGE:
      return {
        ...state,
        code: action.code,
        dirty: true,
      };
    default:
      return state;
  }
};

export const set = (filename, code, ast) => ({
  type: SET,
  filename,
  code: code || null,
  ast: ast || null,
});

export const setAst = ast => ({
  type: SET_AST,
  ast,
});

export const change = code => ({
  type: CHANGE,
  code,
});

export const runParserWithQuery = () => (dispatch, getState) => {
  const state = getState();
  const {
    code: { code, filename },
    options: { customServer, customServerUrl, parseMode, uastQuery },
    languages: { selected: languageSelected },
  } = state;

  dispatch(errorsClear('parse'));
  dispatch({ type: PARSE });

  // there is no action for custom server update, run parser is such an action
  dispatch(versionUpdateIfNeeded());
  dispatch(driversUpdateIfNeeded());

  return api
    .parse(
      parseMode,
      languageSelected,
      filename,
      code,
      uastQuery,
      customServer ? customServerUrl : undefined
    )
    .then(({ uast, language }) => {
      dispatch(languageSet(language));
      dispatch(setAst(uast));
    })
    .catch(errors => {
      dispatch({ type: PARSE_FAILED });
      dispatch(errorsAdd(errors, 'parse'));
    });
};

export const runParser = () => dispatch => {
  dispatch(setUastQuery(''));
  return dispatch(runParserWithQuery());
};
