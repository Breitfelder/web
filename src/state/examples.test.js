import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);

import {
  reducer,
  initialState,
  SET,
  RESET,
  SET_LIST_BY_LANGS,
  select,
} from './examples';
import { set as codeSet } from './code';
import { set as languageSet, SELECT as LANG_SELECT } from './languages';
import { java_example_1 } from '../examples/hello.java.js';

describe('examples/reducer', () => {
  it('SET', () => {
    expect(
      reducer(
        {
          ...initialState,
          list: {
            java_example_1: {
              name: 'hello.java',
              language: 'java',
              code: java_example_1,
            },
          },
        },
        { type: SET, selected: 'java_example_1' }
      )
    ).toMatchSnapshot();
  });

  it('RESET', () => {
    expect(
      reducer(initialState, {
        type: RESET,
      })
    ).toMatchSnapshot();
  });

  it('SET_LIST_BY_LANGS', () => {
    expect(
      reducer(initialState, {
        type: SET_LIST_BY_LANGS,
        languages: ['java', 'python'],
      })
    ).toMatchSnapshot();
  });
});

describe('examples/actions', () => {
  it('select', () => {
    const store = mockStore({
      examples: initialState,
    });

    const key = 'java_example_1';
    const example = {
      name: 'hello.java',
      language: 'java',
      code: java_example_1,
    };
    store.dispatch(select(key));
    expect(store.getActions()).toEqual([
      codeSet(example.name, example.code),
      {
        type: LANG_SELECT,
        selected: '',
      },
      languageSet(example.language),
      {
        type: SET,
        selected: key,
      },
    ]);
  });
});
