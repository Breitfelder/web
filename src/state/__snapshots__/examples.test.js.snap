// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`examples/reducer RESET 1`] = `
Object {
  "list": Object {},
  "selected": "",
}
`;

exports[`examples/reducer SET 1`] = `
Object {
  "list": Object {
    "java_example_1": Object {
      "code": "// hello.java
import java.io.*;
import javax.servlet.*;

public class Hello extends GenericServlet {
    public void service(final ServletRequest request, final ServletResponse response)
    throws ServletException, IOException {
        response.setContentType(\\"text/html\\");
        final PrintWriter pw = response.getWriter();
        try {
            pw.println(\\"Hello, world!\\");
        } finally {
            pw.close();
        }
    }
}
",
      "language": "java",
      "name": "hello.java",
    },
  },
  "selected": "java_example_1",
}
`;

exports[`examples/reducer SET_LIST_BY_LANGS 1`] = `
Object {
  "list": Object {
    "java_example_1": Object {
      "code": "// hello.java
import java.io.*;
import javax.servlet.*;

public class Hello extends GenericServlet {
    public void service(final ServletRequest request, final ServletResponse response)
    throws ServletException, IOException {
        response.setContentType(\\"text/html\\");
        final PrintWriter pw = response.getWriter();
        try {
            pw.println(\\"Hello, world!\\");
        } finally {
            pw.close();
        }
    }
}
",
      "language": "java",
      "name": "hello.java",
    },
    "java_example_2": Object {
      "code": "public class SwapElementsExample {

         public static void main(String[] args) {

                 int num1 = 10;
                 int num2 = 20;

                 System.out.println(\\"Before Swapping\\");
                 System.out.println(\\"Value of num1 is :\\" + num1);
                 System.out.println(\\"Value of num2 is :\\" +num2);

                 //swap the value
                 swap(num1, num2);
         }

         private static void swap(int num1, int num2) {

                 int temp = num1;
                 num1 = num2;
                 num2 = temp;

                 System.out.println(\\"After Swapping\\");
                 System.out.println(\\"Value of num1 is :\\" + num1);
                 System.out.println(\\"Value of num2 is :\\" +num2);

         }
 }
",
      "language": "java",
      "name": "swap.java",
    },
    "python_example_1": Object {
      "code": "def fizzbuzz(n):

    if n % 3 == 0 and n % 5 == 0:
        return 'FizzBuzz'
    elif n % 3 == 0:
        return 'Fizz'
    elif n % 5 == 0:
        return 'Buzz'
    else:
        return str(n)

print \\"\\\\n\\".join(fizzbuzz(n) for n in xrange(1, 21))
",
      "language": "python",
      "name": "fizzbuzz.py",
    },
    "python_example_2": Object {
      "code": "class Animal:
    def __init__(self):
        self._a = 1
        b = 2

    def method(self, arg1):
        pass

    @property
    def a():
        return self._a

    @a.setter
    def a(newa):
        self._a = newa

a = Animal()
a.b = 3
a.b.c.d.e.f.g = 6
a.method(5)
",
      "language": "python",
      "name": "classdef.py",
    },
  },
  "selected": "",
}
`;
